### PR TITLE
Add initial implementation of jax.Array base class

### PR DESCRIPTION
**Pull request by jakevdp**
Part of #12049

The idea here is that `jax.Array` will eventually be defined in C++ and used as a true base class of both `core.Tracer` and `jax.experimental.array.Array`, which will be renamed to `ArrayImpl` or something similar before being moved to `jax._src.array`.

~Still need to add some tests similar to the ones in https://github.com/google/jax/pull/12351 to ensure that this implementation has the desired properties.~ *Done*

### CODE REVIEW

File: jax/_src/array.py [hunk 1]
```diff
@@ -12,12 +12,11 @@
 # See the License for the specific language governing permissions and
 # limitations under the License.
 
+
 import abc
```
Comment (yashk2810) [inline jax/_src/array.py:16]: Are we sure we want to rename this file right now?

I was planning to do it while moving the entire thing in experimental/array.py here and sharding.py too.

Comment (yashk2810) [reply jax/_src/array.py:16]: Currently, now there are 2 types of imports which is confusing. Can we just do the rename when the concrete class moves too?

Comment (jakevdp) [reply jax/_src/array.py:16]: Do you mean we should define `jax.Array` within `jax._src.numpy.ndarray`?

Comment (yashk2810) [reply jax/_src/array.py:16]: Yes, as a stopgap. If you want to.. maybe move the entire Array implementation here with this rename?

But I guess that will touch a lot more files :)

And then you can add a big comment, telling what to move to jax._src.array when you/I do the move.

Comment (jakevdp) [reply jax/_src/array.py:16]: I think I'll put it in `jax._src.typing` for now instead.

File: jax/__init__.py [hunk 2]
```diff
@@ -35,6 +35,8 @@
 from jax import config as _config_module
 del _config_module
 
+from jax._src.basearray import Array as Array
```
Comment (yashk2810) [inline jax/__init__.py:38]: Do we want to export this right now?

File: jax/_src/lax/lax.py [hunk 3]
```diff
@@ -1330,11 +1338,11 @@ def full_like(x: Array, fill_value: ArrayLike, dtype: Optional[DTypeLike] = None
   # probably in the form of a primitive like `val = match_sharding_p.bind(x, val)`
   # (so it works in staged-out code as well as 'eager' code). Related to
   # equi-sharding.
-  if (config.jax_array and hasattr(x, 'sharding') and
-      not dispatch.is_single_device_sharding(x.sharding) and
-      not isinstance(x.sharding, sharding.PmapSharding)):
-    return array.make_array_from_callback(
-        fill_shape, x.sharding, lambda idx: val[idx])  # type: ignore[arg-type]
+  if config.jax_array and hasattr(x, 'sharding'):
+    sharding = x.sharding  # type: ignore[union-attr]
```
Comment (yashk2810) [inline jax/_src/lax/lax.py:1342]: Why this change? Do I have to use this pattern everywhere?

File: jax/__init__.py (hunk 2 above)
Comment (jakevdp) [reply jax/__init__.py:38]: As I understood it, the goal of this was to expose a working prototype of `jax.Array` so we could migrate users to it before landing the new experimental `Array`, rather than having to change their code twice. So yes, we'll need to export this now.

Comment (yashk2810) [reply jax/__init__.py:38]: Ahh right! SGTM!

File: jax/_src/lax/lax.py (hunk 3 above)
Comment (jakevdp) [reply jax/_src/lax/lax.py:1342]: Because `hasattr(x, 'sharding')` is not enough for mypy to understand that `x` is now of a type that has an attribute `sharding`, so `x.sharding` results in a type check error. We could avoid this if every Array-like object declared a `sharding` attribute (e.g. `Tracer` would have to define sharding), or by using an `isinstance` check in place of the `hasattr` check (i.e. avoiding duck-typing) but I did it this way in order to not change the semantics of the function, which I admit I don't totally understand.

Comment (yashk2810) [reply jax/_src/lax/lax.py:1342]: Hmm.. okay. We will eventually end up defining a sharding attributed on Tracer too I think but let's deal with that later.

**Review by yashk2810 — APPROVED** [review]

**Comment by jakevdp** [comment]
I had been making some effort to land independent parts of this separately in order to make the change less extensive (e.g. https://github.com/google/jax/pull/12370). Do you think that's worth doing here? I worry a change like this is big enough that it might cause unexpected issues (e.g. does having `Tracer` inherit from a class without `__slots__` cause performance degredation?) So thought it would be worth landing in small bits. What do you think?

**Comment by yashk2810** [comment]
> does having Tracer inherit from a class without __slots__ cause performance degredation?) 

I think the best way to know this is to write a quick benchmark and see if it regresses. The one good thing I like about this CL is that if it does break something, we can roll it back as 1 CL rather than rolling back parts of it. So I think it's fine to keep this as is.

If you are worried about perf, https://github.com/google/jax/blob/main/benchmarks/api_benchmark.py has been really handy and you can run this locally too!

**Comment by jakevdp** [comment]
I'm more worried about the things I haven't thought of... and changes like #12370 are truly independent – they're correct whether or not we land the rest of this, so there wouldn't have to be rolled back together. There are a few other similar changes in here (e.g. the changes to `lax_numpy.py`)

**Comment by yashk2810** [comment]
Ohh I see. Yeah splitting up SGTM.

File: jax/experimental/array.py [hunk 4]
```diff
@@ -95,7 +95,7 @@ def _reconstruct_array(fun, args, arr_state, aval_state):
 
 
 @pxla.use_cpp_class(xc.Array if xc._version >= 92 else None)
-class Array:
+class Array(basearray.Array):
```
Comment (hawkinsp) [inline jax/experimental/array.py:98]: Note that if the C++ class is in use (i.e., you have a new enough `jaxlib`) then subtyping won't hold for `Array` either.

**Review by hawkinsp — APPROVED** [review]

File: jax/experimental/array.py (hunk 4 above)
Comment (jakevdp) [reply jax/experimental/array.py:98]: Ah, right, so maybe for now we do need `basearray.Array.register(Array)`, which will take care of runtime instance checks. At static type check time, though, I don't think mypy/pytype is smart enough to figure out that the decorator has replaced the class implementation, so I believe it will still work as long as this is the method used to fall-back. What do you think?

Comment (yashk2810) [reply jax/experimental/array.py:99]: Maybe just do the register below even for Python?

File: jax/experimental/array.py [hunk 5]
```diff
@@ -13,6 +13,7 @@
 # limitations under the License.
 
 from __future__ import annotations
+from email.mime import base
```
Comment (yashk2810) [inline jax/experimental/array.py:16]: What's `email.mime`?

Comment (jakevdp) [reply jax/experimental/array.py:16]: Gah, I have to set up vscode to not try to automatically add imports...

I suspect I typed `base` and hit tab, so it helpfully found something named `base` to import for me

File: jax/experimental/array.py (hunk 4 above)
Comment (jakevdp) [reply jax/experimental/array.py:98]: I've done it conditionally based on the jaxlib version, let me know if that looks OK.

Comment (jakevdp) [reply jax/experimental/array.py:98]: No, if we remove the inheritance then mypy/pytype will no longer recognize `Array` as a subclass of `jax.Array`

Comment (yashk2810) [reply jax/experimental/array.py:98]: Ohh I see. Okay, LGTM.

**Comment by mrahtz** [comment]
(Just wanted to say a huge THANK YOU for this Jake - it's something that a lot of us at DeepMind have been wanting for a long time!)
